R/summary_funs.R

```diff
@@ -259,141 +233,178 @@ weighted_summary_means <- function(y_wobs_test, family, wdraws, mu, dis, cl_ref,
+                                   y_idx = loo_inds)
+      value <- srs_diffe$y_hat
+      # combine estimates of var(y_hat) and var(y)
+      value_se <- sqrt(srs_diffe$v_y_hat + srs_diffe$hat_v_y)
+    } else {
+      # full LOO estimator
+      value <- sum(lppd - lppd_baseline, na.rm = TRUE)
+      value_se <-sd(lppd - lppd_baseline, na.rm = TRUE) * sqrt(n_full)
     }
     if (stat %in% c("mlpd", "gmpd")) {
-      value <- value / n_notna
-      value.se <- value.se / n_notna
+      value <- value / n_full
+      value_se <- value_se / n_full
       if (stat == "gmpd") {
         value_gmpd <- exp(value)
-        # Delta method:
-        value_gmpd.se <- value.se * value_gmpd
+        # delta method
+        value_gmpd_se <- value_se * value_gmpd
       }
     }
-  } else if (stat %in% c("mse", "rmse")) {
+  } else if (stat %in% c("mse", "rmse", "R2")) {
     y <- y_wobs_test$y_prop %||% y_wobs_test$y
-    if (!all(y_wobs_test$wobs == 1)) {
-      wcv <- wcv * y_wobs_test$wobs
-      wcv <- n_notna * wcv / sum(wcv)
+    wobs <- y_wobs_test$wobs
+    wobs <- n_full * wobs / sum(wobs)
+    if (!is.null(summaries_baseline)) {
+      mu_baseline <- summaries_baseline$mu
     }
-    if (stat == "mse") {
-      if (!is.null(mu.bs)) {
-        value <- mean(wcv * ((mu - y)^2 - (mu.bs - y)^2), na.rm = TRUE)
-        value.se <- weighted.sd((mu - y)^2 - (mu.bs - y)^2, wcv,
-                                na.rm = TRUE) /
-          sqrt(n_notna)
+    # Use normal approximation for mse and delta method for rmse and R2
+    if (n_loo < n_full) {
+      # subsampling difference estimator (Magnusson et al., 2020)
+      srs_diffe <- .srs_diff_est_w(y_approx = (summaries_fast$mu - y)^2,
+                                   y = ((mu - y)^2)[loo_inds],
+                                   y_idx = loo_inds,
+                                   w = wobs)
+      value <- srs_diffe$y_hat / n_full
+      # combine estimates of var(y_hat) and var(y)
+      value_se <- sqrt(srs_diffe$v_y_hat + srs_diffe$hat_v_y) / n_full
+    } else {
+      # full LOO estimator
+      value <- mean(wobs * (mu - y)^2)
+      value_se <- .weighted_sd((mu - y)^2, wobs) / sqrt(n_full)
+    }
+    # store for later calculations
+    mse_e <- value
+    if (!is.null(summaries_baseline)) {
+      # delta=TRUE, variance of difference of two normally distributed
+      mse_b <- mean(wobs * (mu_baseline - y)^2)
+      var_mse_b <- .weighted_sd((mu_baseline - y)^2, wobs)^2 / n_full
+      if (n_loo < n_full) {
+        mse_e_fast <- mean(wobs * (summaries_fast$mu - y)^2)
+        srs_diffe <-
+          .srs_diff_est_w(y_approx = ((summaries_fast$mu - y)^2 - mse_e_fast) *
+                            ((mu_baseline - y)^2 - mse_b),
+                          y = (((mu - y)^2 - mse_e) *
+                                 ((mu_baseline - y)^2 - mse_b))[loo_inds],
+                          y_idx = loo_inds,
+                          w = wobs)
+        cov_mse_e_b <- srs_diffe$y_hat / n_full^2
```

Comment:
Just for my understanding: This procedure for estimating the covariance between `mse_e` and `mse_b` assumes that the summands within `mse_e` and `mse_b` coming from different observations are uncorrelated, right? At first, I thought this was violated here because `mu`, `mu_baseline`, and `summaries_fast$mu` are model-based and hence there is potential for cross-observations dependencies, but then I realized that `mu`, `mu_baseline`, and `summaries_fast$mu` all are based on the leave-one-out principle, so is this the reason why we can assume a cross-observation correlation of zero here?